shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/LogInViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package co.nimblehq.avishek.phong.kmmic.presentation.module
+
+import co.nimblehq.avishek.phong.kmmic.data.remote.helper.extension.toErrorMessage
+import co.nimblehq.avishek.phong.kmmic.domain.usecase.LogInUseCase
+import co.nimblehq.avishek.phong.kmmic.helper.extension.isValidEmail
+import com.rickclephas.kmp.nativecoroutines.NativeCoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+
+data class LogInViewState(
+    val isSuccess: Boolean = false,
+    val isLoading: Boolean = false,
+    val error: String? = null,
+    val isInvalidEmail: Boolean = false,
+    var isInvalidPassword: Boolean = false
+) {
+    constructor() : this(false)
+
+    constructor(error: String?) : this(false, false, error)
+}
+
+class LogInViewModel(private val logInUseCase: LogInUseCase): BaseViewModel() {
+
+    private val mutableViewState: MutableStateFlow<LogInViewState> =
+        MutableStateFlow(LogInViewState())
+
+    val viewState: StateFlow<LogInViewState> = mutableViewState
+
+    fun logIn(email: String, password: String) {
+        if (!validInput(email, password)) { return }
+        setLoadingState()
+        viewModelScope.launch {
+            logInUseCase(email = email, password = password)
+                .catch { error -> handleLoginError(error) }
+                .collect { loginSuccess() }
+        }
```

Comment:
We can use the `launchIn(viewModelScope)` method to reduce nesting. Also, we can remove the curly braces around `return`.
```suggestion
if (!validInput(email, password)) return
        logInUseCase(email = email, password = password)
            .onStart { setLoadingState() }
            .catch { error -> handleLoginError(error) }
            .onEach { loginSuccess() }
            .launchIn(viewModelScope)
```

Reply:
solved in 27ddae6